core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -244,6 +245,10 @@ public IndexShard(ShardId shardId, IndexSettings indexSettings, ShardPath path,
         suspendableRefContainer = new SuspendableRefContainer();
         searcherWrapper = indexSearcherWrapper;
         primaryTerm = indexSettings.getIndexMetaData().primaryTerm(shardId.id());
+        refreshListeners = new RefreshListeners(
+                indexSettings::getMaxRefreshListeners,
+                () -> refresh("too_many_listeners"),
+                fire -> threadPool.executor(ThreadPool.Names.LISTENER).execute(fire));
```

Comment:
I think we should also just add the RefreshListeners to the EngineConfig and create a new instance of it inside `newEngineConfig`. Then once the engine is created inside `internalPerformTranslogRecovery` we can go and do `engineConfig.getRefreshListeners().setTranslog(engine.getTranslog())` and don't need that extra listener?